rllib/agents/ppo/ddppo.py

```diff
@@ -89,186 +91,186 @@
-                metrics.counters[STEPS_TRAINED_COUNTER] += count
-                metrics.info[LEARNER_INFO] = info
-            # Since SGD happens remotely, the time delay between fetch and
-            # completion is approximately the SGD step time.
-            metrics.timers[LEARN_ON_BATCH_TIMER].push(time.perf_counter() -
-                                                      self.fetch_start_time)
-
-    train_op = (
-        rollouts.for_each(train_torch_distributed_allreduce)  # allreduce
-        .batch_across_shards()  # List[(grad_info, count)]
-        .for_each(RecordStats()))
-
-    train_op = train_op.for_each(update_worker_global_vars)
-
-    # Sync down the weights. As with the sync up, this is not really
-    # needed unless the user is reading the local weights.
-    if config["keep_local_weights_in_sync"]:
-
-        def download_weights(item):
-            workers.local_worker().set_weights(
-                ray.get(workers.remote_workers()[0].get_weights.remote()))
+class DDPPOTrainer(PPOTrainer):
+    @classmethod
+    @override(PPOTrainer)
+    def get_default_config(cls) -> TrainerConfigDict:
+        return DEFAULT_CONFIG
+
+    @override(PPOTrainer)
+    def validate_config(self, config):
+        """Validates the Trainer's config dict.
+
+        Args:
+            config (TrainerConfigDict): The Trainer's config to check.
+
+        Raises:
+            ValueError: In case something is wrong with the config.
+        """
+        # Call (base) PPO's config validation function first.
+        # Note that this will not touch or check on the train_batch_size=-1
+        # setting.
+        super().validate_config(config)
+
+        # Error if run on Win.
+        if sys.platform in ["win32", "cygwin"]:
+            raise ValueError("DD-PPO not supported on Win yet! "
+                             "Due to usage of torch.distributed.")
+
+        # Auto-train_batch_size: Calculate from rollout len and
+        # envs-per-worker.
+        if config["train_batch_size"] == -1:
+            config["train_batch_size"] = (config["rollout_fragment_length"] *
+                                          config["num_envs_per_worker"])
+        # Users should not define `train_batch_size` directly (always -1).
+        else:
+            raise ValueError(
+                "Set rollout_fragment_length instead of train_batch_size "
+                "for DDPPO.")
+
+        # Only supported for PyTorch so far.
+        if config["framework"] != "torch":
+            raise ValueError(
+                "Distributed data parallel is only supported for PyTorch")
+        if config["torch_distributed_backend"] not in ("gloo", "mpi", "nccl"):
+            raise ValueError("Only gloo, mpi, or nccl is supported for "
+                             "the backend of PyTorch distributed.")
+        # `num_gpus` must be 0/None, since all optimization happens on Workers.
+        if config["num_gpus"]:
```

Comment:
I am just curious, if my eval worker runs on head node and needs gpu, which param do I use to configure it?

Reply:
```
config:
   evaluation_config:
      num_gpus_per_worker: ...
```

Reply:
👌